library/alloc/src/rc.rs

```diff
@@ -3360,6 +3382,68 @@ fn data_offset_align(align: usize) -> usize {
     layout.size() + layout.padding_needed_for(align)
 }
 
+/// A unique owning pointer to a [`RcBox`] **that does not imply the contents are initialized,**
+/// but will deallocate it (without dropping the value) when dropped.
+///
+/// This is a helper for [`Rc::make_mut()`] to ensure correct cleanup on panic.
+#[cfg(not(no_global_oom_handling))]
+struct RcUninit<T: ?Sized, A: Allocator> {
```

Comment:
In the meantime we have gotten the `UniqueRc` type, so I think this could be changed to `UniqueRc<MaybeUninit<T>>` instead.

https://doc.rust-lang.org/nightly/std/rc/struct.UniqueRc.html

Reply:
That sounds like an excellent idea — except that `UniqueRc` isn't generic enough to be used here: it has no `A: Allocator` parameter.

Reply:
Looks like UniqueRc predates the the Allocator parameter on Rc, so it could be added now. https://github.com/rust-lang/rust/issues/112566#issuecomment-1614909204

Reply:
Another problem I found while rebasing and updating comments is that there is no parallel `UniqueArc` — which seems like a missing feature itself. Perhaps fixing both of these would make a good separate PR?

Reply:
if you feel that's too much for this pr, sure.

Reply:
Created #126285 for the needed allocator and `?Sized` support. I'll do the duplication to `Arc` in a third PR (or at least another commit) after review, to keep the number of possible accidental desyncs lower.

Reply:
Aaaaaagh. I go to change the code to use `UniqueRc<MaybeUninit<T>>`, and discover it _cannot_ be used for this purpose, because `MaybeUninit` doesn't allow `T: ?Sized` (which support is the entire point of this PR).

I suppose I can rename `RcUninit` to `UniqueRcUninit` to make its role clearer, at least.